packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/subscribe.ts

```diff
@@ -0,0 +1,294 @@
+  }
+
+  public mutateContainerDefinition(props: ecs.ContainerDefinitionOptions): ecs.ContainerDefinitionOptions {
+    return {
+      ...props,
+
+      environment: { ...(props.environment || {}), ...this.environment },
+    } as ecs.ContainerDefinitionOptions;
+  }
+}
+
+/**
+ * This extension creates a default SQS Queue, `eventsQueue`. It also supports creation of SQS Queue
+ * per SNS Topic and sets up the SNS Subscriptions accordingly.
+ *
+ * The created subscriptions can be accessed using `subscriptions` field and the
+ * default queue for this service can be accessed using the getter `<extension>.eventsQueue`.
+ */
+export class SubscribeExtension extends ServiceExtension {
+  /** The subscriptions created by the extension.
+   * @default undefined
+   */
+  public readonly subscriptions?: Subscription[];
+
+  private _eventsQueue!: sqs.IQueue;
+
+  private environment: { [key: string]: string } = {};
+
+  private deadLetterQueues: sqs.IQueue[] = [];
+
+  private props?: SubscribeProps;
+
+  constructor(props?: SubscribeProps) {
+    super('subscribe');
+
+    this.props = props;
+    if (props?.topicSubscriptions) {
+      this.subscriptions = [];
+    }
+
+    if (props?.eventsQueue && props?.eventsQueueProps) {
+      throw Error('You can only specify either a queue or queue props for creating the default events queue.');
+    }
+  }
+
+  /**
+   * This hook creates and sets the default queue `eventsQueue` and dead letter queue and other topic-specific
+   * according to the provided settings. It creates subscriptions for the topics and also adds the queue URLs to the
+   * environment.
+   * @param service The parent service which this extension has been added to
+   * @param scope The scope that this extension should create resources in
+   */
+  public prehook(service: Service, scope: Construct) {
+    this.parentService = service;
+    this.scope = scope;
+
+    if (this.props?.eventsQueue) {
+      this._eventsQueue = this.props.eventsQueue;
+    } else {
+      this._eventsQueue = this.createQueues('Events', this.props?.eventsQueueProps, this.props?.eventsDeadLetterQueueProps);
+    }
+    this.environment[`${this.parentService.id.toUpperCase()}_EVENTS_QUEUE_URL`] = this._eventsQueue.queueUrl;
+
+    if (this.props?.topicSubscriptions) {
+      const eventsQueueSubscriptions = [];
+      for (const topicSubscription of this.props.topicSubscriptions) {
+        const topic = topicSubscription.topic;
+        const topicName = topic.node.id;
+
+        if (topicSubscription.queue && topicSubscription.queueProps) {
+          throw Error(`You can only specify either a queue or queue props for creating a queue for topic ${topicName}.`);
+        }
+
+        if (!topicSubscription.queue && !topicSubscription.queueProps && !topicSubscription.deadLetterQueueProps) {
+          topic.addSubscription(new subscription.SqsSubscription(this._eventsQueue));
+          eventsQueueSubscriptions.push(topic);
+        } else {
```

Comment:
It is confusing that we allow a sub service to be able to define and create its pub...that sounds like an infinite loop to me.

Reply:
Discussed offline and resolving this comment.